Resume downloading/uploading

mixx99 opened the pull request:
## What added

### log.c/log.h

#### Functions

- `void add_json_file(Message *message)`
- `void update_json_file(Message *message)`
- `void remove_json_file(Message *message)`
- `void get_stucked_message(Message *message)`

All these functions are used to work with a json file, namely, adding a file to the queue, updating it, and deleting it. 

For parse JSON files we are using [cJSON](https://github.com/DaveGamble/cJSON).

There is also a feature for pulling out stuck files. So far, it only works with one stuck file in turn(First-come, first-served). Inside there is a delay setting parameter called MAX_DELAY, which is currently set to 3 seconds.

The basic idea is this:
- If we receive a message of the `MESSAGE_FILE_START` type, we add it using function `add_json_file`.
- If we receive a message of the `MESSAGE_FILE_MID` type, we call `update_json_file` function to update the time and packet number.
- At last if we receive a message of the `MESSAGE_FILE_END` we call `remove_json_file` function to remove this file from our json.

And if the difference between current time and `last-update-time` more than `MAX_DELAY` - we will return this object for request a re-upload until it will exist in `files.json`.

### message.c/message.h

There were no major changes in these files, except for optimizing the work for the algorithm described above.

#### Of the interesting features:

The first packet of the file has type `MESSAGE_FILE_START` and fields `Message.name = filename` and `Message.text = path-to-file`.

### User.c/User.h

There are also no major changes here, but some interesting things:

#### Functions

`void *check_download_files(void *user_arg)` - This function has a separate thread, as well as its own `CHECK_STUCKED_FILES_DELAY` parameter, which defaults to 20 seconds, so as not to check or send request messages to the sender of files too often.
`void destroy_user(User *user)` - Sen

ds everyone a farewell message and calls abort.(Of course, it is better to close all sockets and streams than to crash the program, and in general, there is a need to work on the correct termination of the program in emergency cases, possibly in the future...)
`void sigint_handler(int dummy)` - Catch the `SIGINT` signal and calls `destroy_user`.

## Tests

https://github.com/user-attachments/assets/7f92706d-7f9b-4300-b15d-6de07b8cd94f

~Sorry for the quality, it was compressed a few times to make it 10Mb.~

I've tested this locally on various occasions, and it seems to work fine.

mixx99 committed:
resume-download

crc

Andrewbooq commented:
In such complex logic looks like we need some confirmation for downloaded files - CRC or similar. Not sure that I understood about time - check_download_files looks logical, you need to catch the interruption of downloading. Do you pass to the sending side from what place to continue? What if the sys time on machine will be changed or they have different (incorrect) time?

"void send_private_file(User *receiver, User *me, char *path, int packet_number);" Is here "packet_number" place to continue ?

mixx99 commented:
I looked into CRC and this is exactly what we needed. I added a commit in this PR with a CRC check and squashed the commits.

About users having different system times: it doesn’t really affect functionality, since the `message.time` field isn’t used anywhere. If needed, it could be removed without issues. Everywhere we need the time, we just use `time()` locally.

The case where a user’s system time changes during a file transfer is rare and problematic, but not critical. We check for stuck files roughly every 20 seconds, so the chance of hitting the tiny window where a time change would cause us to miss a new file packet and not update the time in the JSON is very low. Of course, rarity doesn’t mean we should ignore it, but for now I think it’s okay to leave it as is.

Regarding the function:
`void send_private_file(User *receiver, User *me, char *path, int packet_number);` - Yes, `packet_number` indicates the packet from which to continue sending.
